### PR TITLE
feat: RayServe with vLLM using AWS Neuron on Amazon EKS

### PR DESCRIPTION
**Pull request by ratnopamc**
### What does this PR do?
Adds capability to deploy LLMs for inference on AWS Inferentia with ray and vLLM
🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

### Motivation
#591 

### More

- [] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [] Mandatory for new blueprints. Yes, I have added a example to support my blueprint PR
- [ ] Mandatory for new blueprints. Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: gen-ai/inference/vllm-rayserve-inf2/ray-service.yaml [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: hf-token
+  namespace: vllm
+data:
+  hf-token: $HUGGING_FACE_HUB_TOKEN
+---
+apiVersion: ray.io/v1
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/ray-service.yaml:17]: I think we want to be a little more consistent with how we set up the resource configuration. We have 3 max replicas. We should be consistent with deciding if we want to scale nodes or if we want to scale actors. If we want to allow for intra node scaling, I believe you want to set the 
```
neuron_codes = (6 * 2) / 3
```
 or 
```
(NUM_NEURON_DEVICES * 2) / NUM_REPLICAS_PER_NODE
```
We also want to make sure we set the `num_cpus = total_node_cpus / num_replicas`

File: gen-ai/inference/vllm-rayserve-inf2/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+RUN cd vllm && git checkout v0.5.0
+COPY patches/vllm_v0.5.0_neuron.patch vllm/vllm_v0.5.0_neuron.patch
+RUN cd vllm && git apply --ignore-space-change --ignore-whitespace vllm_v0.5.0_neuron.patch \
+      && pip3 install -U -r requirements-neuron.txt && pip3 install .
+RUN pip3 install pynvml
+RUN pip3 install ray
+
+WORKDIR /serve_app
+
+COPY vllm_asyncllmengine.py /serve_app/vllm_asyncllmengine.py
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/Dockerfile:73]: What do you think of putting this as a ConfigMap and mounting it onto the head node?

File: gen-ai/inference/vllm-rayserve-inf2/vllm_asyncllmengine.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    prompt: str
+    sampling_params: SamplingParamsModel
+
+@serve.deployment()
+@serve.ingress(app)
+class VLLMDeployment:
+    def __init__(self):
+        engine_args = AsyncEngineArgs(
+            model=os.getenv("MODEL_ID", "TinyLlama/TinyLlama-1.1B-Chat-v1.0"),
+            disable_log_requests=True,
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/vllm_asyncllmengine.py:31]: Can we allow all of these to be gotten from the envvars? That would make this script more generic

File: gen-ai/inference/vllm-rayserve-inf2/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            libarchive-dev \
+            libxml2-dev \
+            libnuma-dev \
+            wget \
+            gnupg2 \
+      && sudo rm -rf /var/lib/apt/lists/*
+
+RUN sudo wget -qO - https://apt.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB > ./GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB
+RUN sudo gpg --no-default-keyring --keyring ./aws_neuron_keyring.gpg --import  ./GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB
+RUN sudo gpg --no-default-keyring --keyring ./aws_neuron_keyring.gpg  --export >  ./aws_neuron.gpg
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/Dockerfile:41]: If you ` > .etc/apt/trusted.gpg.d/` you won't need the next line

File: gen-ai/inference/vllm-rayserve-inf2/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            wget \
+            zlib1g-dev \
+            libarchive-dev \
+            libxml2-dev \
+            libnuma-dev \
+            wget \
+            gnupg2 \
+      && sudo rm -rf /var/lib/apt/lists/*
+
+RUN sudo wget -qO - https://apt.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB > ./GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/Dockerfile:39]: These can be grouped into one layer by having only 1 `RUN` command and using `&& \ ` between all of the commands, like: 
```
RUN sudo wget -qO - https://apt.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB > ./GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB && \
sudo gpg --no-default-keyring --keyring ./aws_neuron_keyring.gpg --import ./GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB && \
sudo gpg --no-default-keyring --keyring ./aws_neuron_keyring.gpg --export > ./.etc/apt/trusted.gpg.d/aws_neuron.gpg && \
sudo rm ./GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB
```

File: gen-ai/inference/vllm-rayserve-inf2/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+RUN sudo gpg --no-default-keyring --keyring ./aws_neuron_keyring.gpg --import  ./GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB
+RUN sudo gpg --no-default-keyring --keyring ./aws_neuron_keyring.gpg  --export >  ./aws_neuron.gpg
+RUN sudo mv ./aws_neuron.gpg /etc/apt/trusted.gpg.d/
+RUN sudo rm ./GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB
+
+RUN sudo add-apt-repository -y  "deb https://apt.repos.neuron.amazonaws.com jammy main"
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/Dockerfile:45]: You can also group all the apt commands together and follow it up with `rm -rf /var/lib/apt/lists/*`

File: gen-ai/inference/vllm-rayserve-inf2/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+RUN sudo apt-get -y update
+
+RUN sudo apt-get -y install aws-neuronx-collectives=2.*
+RUN sudo apt-get -y install aws-neuronx-runtime-lib=2.*
+RUN sudo apt-get -y install aws-neuronx-tools=2.*
+
+RUN pip3 config set global.extra-index-url https://pip.repos.neuron.amazonaws.com
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/Dockerfile:52]: Please group all the pip installs

File: gen-ai/inference/vllm-rayserve-inf2/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+ENV LD_LIBRARY_PATH=/home/ray/anaconda3/lib:$LD_LIBRARY_PATH
+ENV PATH=/opt/aws/neuron/bin:$PATH
+ENV PYTHONDONTWRITEBYTECODE=1
+ENV PYTHONUNBUFFERED=1
+ENV PJRT_DEVICE=NEURON
+
+ENV VLLM_TARGET_DEVICE=neuron
+RUN git clone https://github.com/vllm-project/vllm.git
+RUN cd vllm && git checkout v0.5.0
+COPY patches/vllm_v0.5.0_neuron.patch vllm/vllm_v0.5.0_neuron.patch
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/Dockerfile:65]: If you put the copy ahead of the `RUN`, you can chain all the `RUN`s together

**Review by omrishiv — COMMENTED** [review]
Thanks for adding this! I think it's going to be really helpful. I left some comments, hopefully they make sense.

File: gen-ai/inference/vllm-rayserve-inf2/vllm_asyncllmengine.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+class GenerateRequestModel(BaseModel):
+    prompt: str
+    sampling_params: SamplingParamsModel
+
+@serve.deployment()
+@serve.ingress(app)
+class VLLMDeployment:
+    def __init__(self):
+        engine_args = AsyncEngineArgs(
+            model=os.getenv("MODEL_ID", "TinyLlama/TinyLlama-1.1B-Chat-v1.0"),
```
Comment (askulkarni2) [inline gen-ai/inference/vllm-rayserve-inf2/vllm_asyncllmengine.py:30]: Let's use Llama3 and configure vllm as described here https://gist.github.com/enghwa/dc7c5cf2bfd51751fb626c19facb7f60

File: gen-ai/inference/vllm-rayserve-inf2/Dockerfile (hunk 2 above)
Comment (vara-bonthu) [reply gen-ai/inference/vllm-rayserve-inf2/Dockerfile:73]: Let's modify this to use Open AI API server entrypoint

File: gen-ai/inference/vllm-rayserve-inf2/vllm-rayserve-deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+      maxReplicas: 3
+      rayStartParams:
+        # This setting is critical for inf2/Trn1 node autoscaling with RayServe
+        resources: '"{\"neuron_cores\": 2}"'
+      template:
+        spec:
+          schedulerName: my-scheduler # Correct placement
+          containers:
+          - name: worker
+            image: public.ecr.aws/data-on-eks/vllm-ray-neuron-mistral7b:latest
```
Comment (vara-bonthu) [inline gen-ai/inference/vllm-rayserve-inf2/vllm-rayserve-deployment.yaml:250]: @ratnopamc Update this image with new one

Comment (ratnopamc) [reply gen-ai/inference/vllm-rayserve-inf2/vllm-rayserve-deployment.yaml:250]: Sure, will create a new ecr repo and image.

**Comment by vara-bonthu** [comment]
Next Steps once this PR is merged

1/ Add HF Token to deployment yaml and config map serving script to handle gated models
2/ Website Doc for the deployment

File: gen-ai/inference/vllm-rayserve-inf2/vllm-rayserve-deployment.yaml (hunk 10 above)
Comment (ratnopamc) [reply gen-ai/inference/vllm-rayserve-inf2/vllm-rayserve-deployment.yaml]: @vara-bonthu , Pushed the below changes - 

- Fixed Indentation, whitespacing and formatting of the neuron patch file.
- Created a public repository `vllm-ray2.32.0-inf2-llama3` under data-on-eks and pushed image `vllm-ray2.32.0-inf2-llama3` built with the above patch.
- Updated deployment.yaml to include openai install command.
- Tested autoscaling across nodes, works fine!

Please review.

**Review by vara-bonthu — APPROVED** [review]

File: gen-ai/inference/vllm-rayserve-inf2/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        ca-certificates autoconf automake build-essential git gperf \
+        libre2-dev libssl-dev libtool libcurl4-openssl-dev libb64-dev \
+        libgoogle-perftools-dev patchelf python3-dev python3-pip \
+        python3-setuptools rapidjson-dev scons software-properties-common \
+        pkg-config unzip wget zlib1g-dev libarchive-dev libxml2-dev \
+        libnuma-dev gnupg2 && \
+    rm -rf /var/lib/apt/lists/* && \
+    wget -qO - https://apt.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB | gpg --no-default-keyring --keyring ./aws_neuron_keyring.gpg --import && \
+    gpg --no-default-keyring --keyring ./aws_neuron_keyring.gpg --export > /etc/apt/trusted.gpg.d/aws_neuron.gpg && \
+    add-apt-repository -y "deb https://apt.repos.neuron.amazonaws.com jammy main" && \
```
Comment (omrishiv) [inline gen-ai/inference/vllm-rayserve-inf2/Dockerfile:29]: I think we'd want to separate this into another layer to make it easier to update

Comment (vara-bonthu) [reply gen-ai/inference/vllm-rayserve-inf2/Dockerfile:29]: @ratnopamc Could you please update this in your second PR along with Website doc?

**Pull request merged**